### PR TITLE
watch: can use .magnet files

### PR DESCRIPTION
**Pull request by piec**
Hi,

This patch lets transmission-daemon watch .magnet files too

The patch is based on:
 * https://trac.transmissionbt.com/ticket/4710#comment:18
 * https://forum.transmissionbt.com/viewtopic.php?f=3&t=13377#post_content73143

Cheers,
Pierre

### CODE REVIEW

**Comment by idolpx** [comment]
Why is this PR blocked?  This needs to be merged to master and released.  
Please reconsider it's only a few lines of code and would be a significant benefit to myself and others.
I'm currently using a script to scan for ".magnet" files in my watch folder put there by Jackett that uses RPC to submit them to transmission.  That is a waste of cpu cycles when this could do the trick when it's already watching the folder for files.

**Comment by ckerr** [comment]
@piec could you update this PR to match the `daemon.c` -> `daemon.cc` renaming?

**Comment by piec** [comment]
Yes

**Comment by piec** [comment]
this should do it!

**Comment by piec** [comment]
better now

File: daemon/daemon.cc [hunk 1]
```diff
@@ -228,14 +228,43 @@ static auto onFileAdded(tr_watchdir_t dir, char const* name, void* vsession)
...
+        }
+    }
+    else if (tr_str_has_suffix(name, ".magnet"))
+    {
+        auto content = std::vector<char>{};
+        if (!tr_loadFile(content, filename))
+        {
+            tr_logAddInfo("cannot read file \"%s\"", name);
+            retry = true;
+        }
```
Comment (ckerr) [inline daemon/daemon.cc:255]:
```suggestion
        tr_error* error = nullptr;
        if (!tr_loadFile(content, filename, &error))
        {
            tr_logAddWarn(fmt::format(
                _("Couldn't read '{path}': {error} ({error_code})"),
                fmt::arg("path", handle->path),
                fmt::arg("error", error->message),
                fmt::arg("error_code", error->code)));
            tr_error_free(error);
            retry = true;
        }          
```

File: daemon/daemon.cc [hunk 2]
```diff
@@ -228,14 +228,43 @@ static auto onFileAdded(tr_watchdir_t dir, char const* name, void* vsession)
...
+        else
+        {
+            content.push_back(0); // string terminal 0
```
Comment (ckerr) [inline daemon/daemon.cc:258]:
```suggestion
            content.push_back('\0'); // zero-terminated string
```

File: daemon/daemon.cc [hunk 3]
```diff
@@ -228,14 +228,43 @@ static auto onFileAdded(tr_watchdir_t dir, char const* name, void* vsession)
...
+            auto data = content.data();
```
Comment (ckerr) [inline daemon/daemon.cc:259]:
```suggestion
            auto const* data = std::data(content);
```

File: daemon/daemon.cc [hunk 4]
```diff
@@ -250,7 +279,7 @@ static auto onFileAdded(tr_watchdir_t dir, char const* name, void* vsession)
         bool trash = false;
         bool const test = tr_ctorGetDeleteSource(ctor, &trash);
 
-        if (test && trash)
+      if (test && trash)
```
Comment (ckerr) [inline daemon/daemon.cc:282]:
```suggestion
        if (test && trash)
```

File: daemon/daemon.cc [hunk 5]
```diff
@@ -228,14 +228,43 @@ static auto onFileAdded(tr_watchdir_t dir, char const* name, void* vsession)
...
+    bool retry = false;
+
+    if (tr_str_has_suffix(name, ".torrent"))
+    {
+        if (!tr_ctorSetMetainfoFromFile(ctor, filename.c_str(), nullptr))
+        {
+            retry = true;
+        }
+    }
+    else if (tr_str_has_suffix(name, ".magnet"))
```
Comment (ckerr) [inline daemon/daemon.cc:248]: This is always true due to the test up on line 231

```suggestion
    else // ".magnet" suffix
```

**Comment by piec** [comment]
ok @ckerr I'm taking a look at that

**Comment by piec** [comment]
Thanks for the review, I've applied the suggestions, squashed and rebased the changes on main.

**Review by ckerr — APPROVED** [review]
Thanks!

**Comment by GScottElshaw** [comment]
Is this something that can be added to the main client?

**Comment by ckerr** [comment]
CI failure is unrelated to PR

**Pull request merged**